### PR TITLE
READY : update "-f" with using notify_debouncer_full for crossplatform build

### PR DESCRIPTION
**Pull request by SupperZum**

### CODE REVIEW

File: text/src/tail.rs [hunk 1]
```diff
@@ -149,53 +177,56 @@ fn tail(args: &Args) -> Result<(), Box<dyn std::error::Error>> {
...
+
+                                io::stdout().flush()?;
+
+                                last_known_contents = new_contents;
+                                last_known_hash = new_hash;
+                            }
+                        }
+                        EventKind::Remove(RemoveKind::File) => {
+                            eprintln!("File {} deleted, exiting...", file_path.display());
+                            std::process::exit(1);
```
Comment (0x07C0) [inline text/src/tail.rs:222]: Don't print anything or exit. Just remove the watcher

File: text/Cargo.toml [hunk 2]
```diff
@@ -15,7 +15,7 @@ diff = "0.1"
 regex.workspace = true
 chrono.workspace = true
 libc.workspace = true
-inotify = "0.10.2"
+notify-debouncer-full = "0.3.1"
```
Comment (0x07C0) [inline text/Cargo.toml:18]: What's the benefit over simply `notify` crate?

File: text/src/tail.rs [hunk 3]
```diff
@@ -149,53 +177,56 @@ fn tail(args: &Args) -> Result<(), Box<dyn std::error::Error>> {
...
+
+                            if new_hash != last_known_hash {
+                                // If the file content is changed not only by adding new data
+                                if new_contents.starts_with(&last_known_contents) {
+                                    let new_data = &new_contents[last_known_contents.len()..];
+
+                                    print_bytes(new_data);
+                                } else {
+                                    eprintln!("\ntail: {}: file truncated", file_path.display());
+                                    print_bytes(&new_contents);
```
Comment (0x07C0) [inline text/src/tail.rs:211]: `if .. else` could return bytes to print instead of duplicating the print

File: text/src/tail.rs [hunk 4]
```diff
@@ -149,53 +177,56 @@ fn tail(args: &Args) -> Result<(), Box<dyn std::error::Error>> {
...
+        let mut debouncer = new_debouncer(Duration::from_millis(1), None, tx).unwrap();
+
+        // Add a path to be watched.
+        // below will be monitored for changes.
+        debouncer
+            .watcher()
+            .watch(Path::new(file_path), RecursiveMode::NonRecursive)?;
+
+        for res in rx {
+            match res {
```
Comment (0x07C0) [inline text/src/tail.rs:195]: maybe could be flattened?

File: text/src/tail.rs [hunk 5]
```diff
@@ -149,53 +177,56 @@ fn tail(args: &Args) -> Result<(), Box<dyn std::error::Error>> {
...
+                Ok(events) => {
+                    let event = events.first().unwrap();
+                    match event.kind {
+                        EventKind::Modify(_) => {
```
Comment (0x07C0) [inline text/src/tail.rs:199]: We don't need all modify events. only data would be enough I think

File: text/Cargo.toml (hunk 2 above)
Comment (SupperZum) [reply text/Cargo.toml:18]: - Only emits a single Rename event if the rename From and To events can be matched
- Merges multiple Rename events
- Takes Rename events into account and updates paths for events that occurred before the rename event, but which haven’t been emitted, yet
- Optionally keeps track of the file system IDs all files and stiches rename events together (FSevents, Windows)
- Emits only one Remove event when deleting a directory (inotify)
- Doesn’t emit duplicate create events
-Doesn’t emit Modify events after a Create event

**Review by 0x07C0 — APPROVED** [review]

File: text/src/tail.rs [hunk 6]
```diff
@@ -121,6 +125,30 @@ fn print_last_n_bytes<R: Read>(buf_reader: &mut R, n: isize) -> Result<(), Strin
...
+    t.hash(&mut s);
+    s.finish()
+}
+
+fn read_file_to_bytes<P: AsRef<Path>>(file_path: P) -> io::Result<Vec<u8>> {
+    let mut file = File::open(file_path)?;
+    let mut contents = vec![];
+    file.read_to_end(&mut contents)?;
+    Ok(contents)
+}
```
Comment (0x07C0) [inline text/src/tail.rs:139]: Don't read the whole file

File: text/src/tail.rs [hunk 7]
```diff
@@ -121,6 +125,30 @@ fn print_last_n_bytes<R: Read>(buf_reader: &mut R, n: isize) -> Result<(), Strin
...
 }
 
+fn calculate_hash<T: Hash>(t: &T) -> u64 {
+    let mut s = DefaultHasher::new();
+    t.hash(&mut s);
+    s.finish()
+}
+
+fn read_file_to_bytes<P: AsRef<Path>>(file_path: P) -> io::Result<Vec<u8>> {
+    let mut file = File::open(file_path)?;
```
Comment (0x07C0) [inline text/src/tail.rs:135]: Don't reopen on each modify.

File: text/src/tail.rs [hunk 8]
```diff
@@ -149,53 +177,57 @@ fn tail(args: &Args) -> Result<(), Box<dyn std::error::Error>> {
...
-                    }
-
-                    // Read the new lines and output them
-                    let mut new_data = String::new();
-                    let bytes_read = reader.read_to_string(&mut new_data)?;
-                    if bytes_read > 0 {
-                        print!("{}", new_data);
-                        io::stdout().flush()?;
+        let mut last_known_contents = read_file_to_bytes(file_path)?;
+        let mut last_known_hash = calculate_hash(&last_known_contents);
```
Comment (0x07C0) [inline text/src/tail.rs:182]: No need in hashing or storing the whole file in memory. This would fill the whole memory with large files and crash

**Pull request merged**